[Image Set]: Refactor CSSImageSetValue to accept optional resolution and MIME type parameters

rreno opened the pull request:
#### f9b1ad4c0847a6799049abdb764a5ca7367b4800
<pre>
[Image Set]: Refactor CSSImageSetValue to accept optional resolution and MIME type parameters
<a href="https://bugs.webkit.org/show_bug.cgi?id=252137">https://bugs.webkit.org/show_bug.cgi?id=252137</a>
rdar://105367742

Reviewed by Antti Koivisto.

The standardized image-set differs from -webkit-image-set in that each option has three arguments instead of two -
an image, an optional resolution, and an optional type() function declaring the MIME type of the image.

This refactors the CSSImageSetValue to hold an CSSImageSetOptionValue object instead of placing the image and
resolution values in a vector and using a stride of 2 to represent each option. The actual parsing of
resolution and type() will be in a follow-on patch.

* LayoutTests/TestExpectations:
    Unskip newly passing WPT ref tests.

* LayoutTests/fast/css/image-set-parsing-expected.txt:
* LayoutTests/fast/css/image-set-parsing-generated-expected.txt: Removed.
* LayoutTests/fast/css/image-set-parsing-generated.html: Removed.
* LayoutTests/fast/css/image-set-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-parsing.html:
    Refactor the internal parsing tests in the following ways:
      1) Move and dedup all tests of standard properties (e.g. background-image) to WPT.
      2) Move generated image tests (gradients, for example) to the main image-set-parsing test.
      3) Test that -webkit-image-set is an alias to image-set
      4) Test that the serialization is as expected (per WPT).

* Source/WebCore/css/CSSImageSetOptionValue.cpp: Added.
(WebCore::CSSImageSetOptionValue::Type::Type):
(WebCore::CSSImageSetOptionValue::Type::cssText const):
(WebCore::CSSImageSetOptionValue::CSSImageSetOptionValue):
(WebCore::CSSImageSetOptionValue::create):
(WebCore::CSSImageSetOptionValue::equals const):
(WebCore::CSSImageSetOptionValue::customCSSText const):
(WebCore::CSSImageSetOptionValue::image const):
(WebCore::CSSImageSetOptionValue::resolution const):
(WebCore::CSSImageSetOptionValue::setResolution):
(WebCore::CSSImageSetOptionValue::type const):
(WebCore::CSSImageSetOptionValue::setType):
* Source/WebCore/css/CSSImageSetOptionValue.h: Copied from Source/WebCore/css/CSSImageSetValue.h.
    Create a new CSSImageSetOptionValue class to represent the arguments to the image-set function.
    This object holds a Ref to either a CSSImageValue or a generated image (e.g. a linear-gradient),
    and optionally holds a resolution (scale factor) and a MIME type string.

* Source/WebCore/css/CSSImageSetValue.cpp:
(WebCore::CSSImageSetValue::create):
(WebCore::CSSImageSetValue::CSSImageSetValue):
(WebCore::CSSImageSetValue::customCSSText const):
(WebCore::CSSImageSetValue::createStyleImage const):
* Source/WebCore/css/CSSImageSetValue.h:
    With -webkit-image-set, the image and resolution were the only two arguments and they were both
    required. The standardized image-set makes resolution and type optional. Instead of holding all
    the arguments as individual elements in the underlying CSSListValue storage we hold the new
    CSSImageSetOptionValues in the internal vector instead. We also now filter the image set on provided
    MIME type if it&apos;s supported or not when creating the StyleImage.

* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isImageSetOptionValue const):
    Updates for support for the new CSSImageSetOptionValue.

* Source/WebCore/css/calc/CSSCalcPrimitiveValueNode.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/CSSValuePair.cpp:
    Adding a new file caused build errors due to missing includes/incomplete types from forward decls.
    I added the needed includes.

(WebCore::CSSPropertyParserHelpers::consumeImageSetOptionalArguments):
(WebCore::CSSPropertyParserHelpers::consumeImageSet):
    Use the new CSSImageSetOptionValue/CSSImageSetValue API when constructing those objects. Also created
    a stub function which will eventually dispatch to parsing the resolution and type() optional parameters
    in a forthcoming patch.

* Source/WebCore/rendering/style/StyleImageSet.cpp:
(WebCore::StyleImageSet::computedStyleValue const):

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/260760@main">https://commits.webkit.org/260760@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e42e7e6f256530da5616419beb5244e161936ff8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109298 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18378 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/42095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113180 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/19827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/9668 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101539 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115053 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/19827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/42095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/19827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/42095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11152 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/42095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/11870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/9668 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17261 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/42095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7426 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13504 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rreno commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -4417,28 +4420,49 @@ static RefPtr<CSSValue> consumeGeneratedImage(CSSParserTokenRange& range, const
     return result;
 }
 
+static auto consumeImageSetOptionalArguments(CSSParserTokenRange& range)
```
Thinking about this approach a bit more, I don't really like it as much.

I think for the parsing I'm going to try to adopt the Meta Consumer objects that are elsewhere in this file. Something like
```C++
struct ImageSetOptionalArgumentConsumer {
    using FunctionToken = ImageSetOptionalTypeArgumentConsumer;
    using DimensionToken = ResolutionCSSPrimitive(...)Consumer;
};
```

anttijk commented on Source/WebCore/css/CSSValuePair.cpp:
```diff
@@ -26,6 +26,8 @@
 #include "config.h"
 #include "CSSValuePair.h"
 
+#include <wtf/text/WTFString.h>
```
unnecessary empty line

rreno committed:
[Image Set]: Refactor CSSImageSetValue to accept optional resolution and MIME type parameters
https://bugs.webkit.org/show_bug.cgi?id=252137
rdar://105367742

Reviewed by Antti Koivisto.

The standardized image-set differs from -webkit-image-set in that each option has three arguments instead of two -
an image, an optional resolution, and an optional type() function declaring the MIME type of the image.

This refactors the CSSImageSetValue to hold an CSSImageSetOptionValue object instead of placing the image and
resolution values in a vector and using a stride of 2 to represent each option. The actual parsing of
resolution and type() will be in a follow-on patch.

* LayoutTests/TestExpectations:
    Unskip newly passing WPT ref tests.

* LayoutTests/fast/css/image-set-parsing-expected.txt:
* LayoutTests/fast/css/image-set-parsing-generated-expected.txt: Removed.
* LayoutTests/fast/css/image-set-parsing-generated.html: Removed.
* LayoutTests/fast/css/image-set-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-parsing.html:
    Refactor the internal parsing tests in the following ways:
      1) Move and dedup all tests of standard properties (e.g. background-image) to WPT.
      2) Move generated image tests (gradients, for example) to the main image-set-parsing test.
      3) Test that -webkit-image-set is an alias to image-set
      4) Test that the serialization is as expected (per WPT).

* Source/WebCore/css/CSSImageSetOptionValue.cpp: Added.
(WebCore::CSSImageSetOptionValue::Type::Type):
(WebCore::CSSImageSetOptionValue::Type::cssText const):
(WebCore::CSSImageSetOptionValue::CSSImageSetOptionValue):
(WebCore::CSSImageSetOptionValue::create):
(WebCore::CSSImageSetOptionValue::equals const):
(WebCore::CSSImageSetOptionValue::customCSSText const):
(WebCore::CSSImageSetOptionValue::image const):
(WebCore::CSSImageSetOptionValue::resolution const):
(WebCore::CSSImageSetOptionValue::setResolution):
(WebCore::CSSImageSetOptionValue::type const):
(WebCore::CSSImageSetOptionValue::setType):
* Source/WebCore/css/CSSImageSetOptionValue.h: Copied from Source/WebCore/css/CSSImageSetValue.h.
    Create a new CSSImageSetOptionValue class to represent the arguments to the image-set function.
    This object holds a Ref to either a CSSImageValue or a generated image (e.g. a linear-gradient),
    and optionally holds a resolution (scale factor) and a MIME type string.

* Source/WebCore/css/CSSImageSetValue.cpp:
(WebCore::CSSImageSetValue::create):
(WebCore::CSSImageSetValue::CSSImageSetValue):
(WebCore::CSSImageSetValue::customCSSText const):
(WebCore::CSSImageSetValue::createStyleImage const):
* Source/WebCore/css/CSSImageSetValue.h:
    With -webkit-image-set, the image and resolution were the only two arguments and they were both
    required. The standardized image-set makes resolution and type optional. Instead of holding all
    the arguments as individual elements in the underlying CSSListValue storage we hold the new
    CSSImageSetOptionValues in the internal vector instead. We also now filter the image set on provided
    MIME type if it's supported or not when creating the StyleImage.

* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isImageSetOptionValue const):
    Updates for support for the new CSSImageSetOptionValue.

* Source/WebCore/css/calc/CSSCalcPrimitiveValueNode.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/CSSValuePair.cpp:
    Adding a new file caused build errors due to missing includes/incomplete types from forward decls.
    I added the needed includes.

(WebCore::CSSPropertyParserHelpers::consumeImageSetOptionalArguments):
(WebCore::CSSPropertyParserHelpers::consumeImageSet):
    Use the new CSSImageSetOptionValue/CSSImageSetValue API when constructing those objects. Also created
    a stub function which will eventually dispatch to parsing the resolution and type() optional parameters
    in a forthcoming patch.

* Source/WebCore/rendering/style/StyleImageSet.cpp:
(WebCore::StyleImageSet::computedStyleValue const):

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/260760@main
